umake/frameworks/swift.py

```diff
@@ -0,0 +1,120 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2014 Canonical
+#
+# Authors:
+#  Didier Roche
+#
+# This program is free software; you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the Free Software
+# Foundation; version 3.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+
+"""Swift module"""
+
+from contextlib import suppress
+from gettext import gettext as _
+import logging
+import os
+import re
+import lsb_release
+import gnupg
+
+import umake.frameworks.baseinstaller
+from umake.interactions import DisplayMessage
+from umake.tools import get_current_arch, add_env_to_user, ChecksumType, MainLoop
+from umake.network.download_center import DownloadCenter, DownloadItem
+from umake.ui import UI
+
+logger = logging.getLogger(__name__)
+
+
+class SwiftCategory(umake.frameworks.BaseCategory):
+
+    def __init__(self):
+        super().__init__(name="Swift", description=_("Swift language"),
+                         logo_path=None)
+
+
+class SwiftLang(umake.frameworks.baseinstaller.BaseInstaller):
+
+    def __init__(self, category):
+        super().__init__(name="Swift Lang", description=_("Swift compiler (default)"), is_category_default=True,
+                         packages_requirements=["clang", "python3-gnupg"],
+                         category=category, only_on_archs=['amd64'],
+                         download_page="https://swift.org/download/",
+                         dir_to_decompress_in_tarball="swift*",
+                         required_files_path=[os.path.join("usr", "bin", "swift")])
+        self.release = lsb_release.get_distro_information()['RELEASE']
+
+    def parse_download_link(self, line, in_download):
+        """Parse Swift download link, expect to find a .sig file"""
+        url_found = False
+        if self.release in line and '.tar.gz.sig' in line:
+            in_download = True
+        else:
+            in_download = False
+        if in_download:
+            p = re.search(r'href="(.*)" title="PGP Signature"', line)
+            with suppress(AttributeError):
+                self.sig_url = "https://swift.org" + p.group(1)
+                logger.debug(self.sig_url)
+                url_found = True
+                DownloadCenter(urls=[DownloadItem(self.sig_url, None)],
```

Comment:
In case you do the more advanced logic, it means that you need to move starting the download at line 94 (once the whole file is parsing) under a "if url_found".